anndata/tests/test_readwrite.py

```diff
@@ -405,6 +406,34 @@ def test_write_categorical(tmp_path):
     assert np.all(orig.obs.stack().dropna() == curr.obs.stack().dropna())
 
 
+def test_write_large_categorical(tmp_path, diskfmt):
+    M = int(1e5)
+    N = 1000
+    ls = np.array(list(ascii_letters))
+    cats = np.array(
+        sorted(
+            "".join(np.random.choice(ls, np.random.choice(range(5, 30))))
+            for i in range(10000)
```

Comment:
How long does this one take? (I guess we should run `pytest -v`)

Reply:
`pytest --durations=n` reports the top n times. Each test takes ~2.7-2.8 seconds on my machine. About .7 seconds of that is data generation.

If you're interested, here's the result of `pytest --durations=0`:

<details>
<summary>Test timings </summary>

```
====================================== slowest test durations ======================================
2.73s call     anndata/tests/test_readwrite.py::test_write_large_categorical[h5ad]
2.70s call     anndata/tests/test_readwrite.py::test_write_large_categorical[zarr]
1.00s call     anndata/tests/test_readwrite.py::test_readwrite_equivolent_h5ad_zarr[typ0]
0.84s call     anndata/tests/test_readwrite.py::test_readwrite_equivolent_h5ad_zarr[csc_matrix]
0.81s call     anndata/tests/test_readwrite.py::test_readwrite_equivolent_h5ad_zarr[csr_matrix]
0.66s call     anndata/tests/test_readwrite.py::test_zarr_chunk_X
0.59s call     anndata/tests/test_readwrite.py::test_scanpy_pbmc68k[h5ad]
0.35s call     anndata/tests/test_readwrite.py::test_maintain_layers
0.29s call     anndata/tests/test_readwrite.py::test_read_write_maintain_obsmvarm_dtypes
0.28s call     anndata/tests/test_readwrite.py::test_scanpy_pbmc68k[zarr]
0.24s call     anndata/tests/test_hdf5_backing.py::test_backed_raw_subset[array_subset-array_subset]
0.24s call     anndata/tests/test_hdf5_backing.py::test_backed_raw_subset[array_int_subset-array_subset]
0.24s call     anndata/tests/test_hdf5_backing.py::test_backed_raw_subset[array_subset-slice_subset]
0.24s call     anndata/tests/test_hdf5_backing.py::test_backed_raw_subset[array_subset-array_bool_subset]
0.23s call     anndata/tests/test_hdf5_backing.py::test_backed_raw_subset[array_int_subset-slice_subset]
0.23s call     anndata/tests/test_hdf5_backing.py::test_backed_raw_subset[array_bool_subset-array_subset]
0.23s call     anndata/tests/test_hdf5_backing.py::test_backed_raw_subset[slice_subset-array_int_subset]
0.23s call     anndata/tests/test_hdf5_backing.py::test_backed_raw_subset[array_int_subset-array_bool_subset]
0.23s call     anndata/tests/test_hdf5_backing.py::test_backed_raw_subset[array_subset-array_int_subset]
0.23s call     anndata/tests/test_hdf5_backing.py::test_backed_raw_subset[array_bool_subset-array_bool_subset]
0.23s call     anndata/tests/test_hdf5_backing.py::test_backed_raw_subset[array_bool_subset-array_int_subset]
0.23s call     anndata/tests/test_hdf5_backing.py::test_backed_raw_subset[single_subset-array_int_subset]
0.23s call     anndata/tests/test_hdf5_backing.py::test_backed_raw_subset[slice_subset-array_bool_subset]
0.22s call     anndata/tests/test_hdf5_backing.py::test_backed_raw_subset[array_int_subset-array_int_subset]
0.22s call     anndata/tests/test_hdf5_backing.py::test_backed_raw_subset[array_subset-single_subset]
0.22s call     anndata/tests/test_hdf5_backing.py::test_backed_raw_subset[single_subset-slice_subset]
0.22s call     anndata/tests/test_hdf5_backing.py::test_backed_raw_subset[array_bool_subset-single_subset]
0.22s call     anndata/tests/test_hdf5_backing.py::test_backed_raw_subset[slice_subset-slice_subset]
0.22s call     anndata/tests/test_hdf5_backing.py::test_backed_raw_subset[single_subset-array_bool_subset]
0.22s call     anndata/tests/test_hdf5_backing.py::test_backed_raw_subset[array_int_subset-single_subset]
0.22s call     anndata/tests/test_hdf5_backing.py::test_backed_raw_subset[array_bool_subset-slice_subset]
0.22s call     anndata/tests/test_hdf5_backing.py::test_backed_raw_subset[slice_subset-array_subset]
0.22s call     anndata/tests/test_hdf5_backing.py::test_backed_raw_subset[single_subset-array_subset]
0.21s call     anndata/tests/test_hdf5_backing.py::test_backed_raw_subset[slice_subset-single_subset]
0.21s call     anndata/tests/test_hdf5_backing.py::test_backed_raw
0.21s call     anndata/tests/test_hdf5_backing.py::test_backed_raw_subset[single_subset-single_subset]
0.21s call     anndata/tests/test_readwrite.py::test_readwrite_roundtrip[zarr-zarr-csr_matrix]
0.20s call     anndata/tests/test_readwrite.py::test_changed_obs_var_names[zarr]
0.18s call     anndata/tests/test_readwrite.py::test_readwrite_zarr[csr_matrix]
0.17s call     anndata/tests/test_readwrite.py::test_readwrite_zarr[typ0]
0.15s call     anndata/tests/test_readwrite.py::test_readwrite_roundtrip[zarr-zarr-typ0]
0.14s call     anndata/tests/test_views.py::test_view_of_view[np_array-array_int_subset-array_int_subset]
0.14s call     anndata/tests/test_views.py::test_view_of_view[scipy_csc-array_int_subset-slice_subset]
0.13s call     anndata/tests/test_readwrite.py::test_readwrite_roundtrip[h5ad-zarr-csr_matrix]
0.12s call     anndata/tests/test_readwrite.py::test_readwrite_h5ad[dataset_kwargs1-csr_matrix]
0.12s call     anndata/tests/test_readwrite.py::test_readwrite_roundtrip[h5ad-zarr-typ0]
0.12s call     anndata/tests/test_readwrite.py::test_readwrite_roundtrip[zarr-h5ad-typ0]
0.12s call     anndata/tests/test_readwrite.py::test_readwrite_roundtrip[zarr-h5ad-csr_matrix]
0.11s call     anndata/tests/test_readwrite.py::test_readwrite_h5ad[dataset_kwargs0-csr_matrix]
0.11s call     anndata/tests/test_readwrite.py::test_backwards_compat_zarr
0.11s call     anndata/tests/test_views.py::test_view_of_view[np_array-array_int_subset-array_bool_subset]
0.11s call     anndata/tests/test_views.py::test_view_of_view[np_array-array_bool_subset-slice_subset]
0.10s call     anndata/tests/test_views.py::test_view_delitem[varm]
0.10s call     anndata/tests/test_readwrite.py::test_changed_obs_var_names[h5ad]
0.10s call     anndata/tests/test_readwrite.py::test_readwrite_h5ad[dataset_kwargs1-typ0]
0.10s call     anndata/tests/test_views.py::test_view_of_view[scipy_csr-array_int_subset-single_subset]
0.10s call     anndata/tests/test_views.py::test_view_failed_delitem[obsp]
0.10s call     anndata/tests/test_views.py::test_view_failed_delitem[obsm]
0.10s call     anndata/tests/test_views.py::test_view_delitem[obsp]
0.10s call     anndata/tests/test_views.py::test_view_failed_delitem[varm]
0.10s call     anndata/tests/test_views.py::test_view_of_view[scipy_csc-array_bool_subset-array_bool_subset]
0.10s call     anndata/tests/test_inplace_subset.py::test_inplace_subset_obs[scipy_csc-slice_subset]
0.10s call     anndata/tests/test_readwrite.py::test_readwrite_h5ad[dataset_kwargs0-typ0]
0.10s call     anndata/tests/test_views.py::test_view_delitem[varp]
0.10s call     anndata/tests/test_views.py::test_view_delitem[obsm]
0.10s call     anndata/tests/test_views.py::test_view_failed_delitem[varp]
0.09s call     anndata/tests/test_views.py::test_view_failed_delitem[layers]
0.09s call     anndata/tests/test_views.py::test_view_of_view[np_array-array_bool_subset-array_subset]
0.09s call     anndata/tests/test_helpers.py::test_assert_equal
0.09s call     anndata/tests/test_views.py::test_view_delitem[layers]
0.09s call     anndata/tests/test_views.py::test_view_of_view[scipy_csr-array_bool_subset-single_subset]
0.09s call     anndata/tests/test_views.py::test_view_of_view[scipy_csr-array_bool_subset-array_bool_subset]
0.09s call     anndata/tests/test_views.py::test_view_of_view[np_array-array_int_subset-slice_subset]
0.09s call     anndata/tests/test_views.py::test_view_of_view[np_array-array_subset-single_subset]
0.09s call     anndata/tests/test_views.py::test_view_of_view[scipy_csr-array_bool_subset-array_subset]
0.09s call     anndata/tests/test_views.py::test_view_of_view[scipy_csr-array_subset-slice_subset]
0.09s call     anndata/tests/test_views.py::test_view_of_view[scipy_csr-array_bool_subset-slice_subset]
0.09s call     anndata/tests/test_views.py::test_view_of_view[scipy_csc-array_subset-slice_subset]
0.09s call     anndata/tests/test_views.py::test_view_of_view[scipy_csc-array_bool_subset-slice_subset]
0.09s call     anndata/tests/test_views.py::test_view_of_view[np_array-array_bool_subset-array_bool_subset]
0.09s call     anndata/tests/test_views.py::test_view_of_view[np_array-array_subset-array_subset]
0.09s call     anndata/tests/test_views.py::test_view_of_view[scipy_csc-array_subset-single_subset]
0.09s call     anndata/tests/test_views.py::test_view_of_view[scipy_csr-array_int_subset-array_subset]
0.09s call     anndata/tests/test_views.py::test_view_of_view[scipy_csr-array_int_subset-slice_subset]
0.09s call     anndata/tests/test_views.py::test_view_of_view[scipy_csr-array_int_subset-array_bool_subset]
0.09s call     anndata/tests/test_views.py::test_view_of_view[scipy_csr-array_subset-array_bool_subset]
0.09s call     anndata/tests/test_views.py::test_view_of_view[scipy_csr-array_subset-array_int_subset]
0.09s call     anndata/tests/test_views.py::test_view_of_view[scipy_csc-array_int_subset-array_int_subset]
0.09s call     anndata/tests/test_views.py::test_view_of_view[scipy_csc-array_int_subset-array_subset]
0.09s call     anndata/tests/test_views.py::test_view_of_view[scipy_csc-array_subset-array_subset]
0.09s call     anndata/tests/test_views.py::test_view_of_view[np_array-array_subset-slice_subset]
0.09s call     anndata/tests/test_views.py::test_view_of_view[scipy_csc-array_bool_subset-single_subset]
0.09s call     anndata/tests/test_views.py::test_view_of_view[scipy_csc-array_bool_subset-array_subset]
0.09s call     anndata/tests/test_views.py::test_view_of_view[scipy_csr-array_subset-array_subset]
0.09s call     anndata/tests/test_views.py::test_view_of_view[np_array-array_bool_subset-array_int_subset]
0.09s call     anndata/tests/test_views.py::test_view_of_view[scipy_csr-array_int_subset-array_int_subset]
0.09s call     anndata/tests/test_views.py::test_view_of_view[scipy_csr-slice_subset-array_bool_subset]
0.09s call     anndata/tests/test_views.py::test_view_of_view[scipy_csc-array_subset-array_bool_subset]
0.09s call     anndata/tests/test_views.py::test_view_of_view[scipy_csr-array_bool_subset-array_int_subset]
0.09s call     anndata/tests/test_views.py::test_view_of_view[scipy_csr-array_subset-single_subset]
0.09s call     anndata/tests/test_views.py::test_view_of_view[scipy_csc-array_subset-array_int_subset]
0.09s call     anndata/tests/test_views.py::test_view_of_view[np_array-array_subset-array_bool_subset]
0.09s call     anndata/tests/test_views.py::test_view_of_view[scipy_csc-array_int_subset-array_bool_subset]
0.08s call     anndata/tests/test_views.py::test_view_of_view[np_array-array_int_subset-array_subset]
0.08s call     anndata/tests/test_views.py::test_view_of_view[scipy_csc-array_bool_subset-array_int_subset]
0.08s call     anndata/tests/test_views.py::test_view_of_view[scipy_csc-array_int_subset-single_subset]
0.08s call     anndata/tests/test_views.py::test_view_of_view[np_array-array_int_subset-single_subset]
0.08s call     anndata/tests/test_views.py::test_view_of_view[scipy_csc-slice_subset-array_bool_subset]
0.08s call     anndata/tests/test_views.py::test_view_of_view[np_array-array_subset-array_int_subset]
0.08s call     anndata/tests/test_views.py::test_view_of_view[np_array-array_bool_subset-single_subset]
0.08s call     anndata/tests/test_readwrite.py::test_readwrite_roundtrip[h5ad-h5ad-csr_matrix]
0.08s call     anndata/tests/test_views.py::test_view_of_view[scipy_csc-slice_subset-array_int_subset]
0.08s call     anndata/tests/test_views.py::test_view_of_view[np_array-slice_subset-array_int_subset]
0.08s call     anndata/tests/test_readwrite.py::test_readwrite_roundtrip[h5ad-h5ad-typ0]
0.08s call     anndata/tests/test_views.py::test_view_of_view[scipy_csr-slice_subset-array_subset]
0.08s call     anndata/tests/test_views.py::test_view_of_view[scipy_csr-slice_subset-single_subset]
0.08s call     anndata/tests/test_hdf5_backing.py::test_backing
0.08s call     anndata/tests/test_views.py::test_view_of_view[scipy_csc-slice_subset-array_subset]
0.08s call     anndata/tests/test_views.py::test_view_of_view[scipy_csr-slice_subset-array_int_subset]
0.08s call     anndata/tests/test_views.py::test_view_of_view[np_array-slice_subset-array_bool_subset]
0.08s call     anndata/tests/test_views.py::test_view_of_view[scipy_csc-slice_subset-single_subset]
0.08s call     anndata/tests/test_inplace_subset.py::test_inplace_subset_obs[np_array-array_int_subset]
0.08s call     anndata/tests/test_views.py::test_view_of_view[np_array-slice_subset-single_subset]
0.07s call     anndata/tests/test_views.py::test_view_of_view[scipy_csr-slice_subset-slice_subset]
0.07s call     anndata/tests/test_views.py::test_view_of_view[np_array-slice_subset-array_subset]
0.07s call     anndata/tests/test_views.py::test_view_of_view[scipy_csc-slice_subset-slice_subset]
0.07s call     anndata/tests/test_inplace_subset.py::test_inplace_subset_var[scipy_csc-array_int_subset]
0.07s call     anndata/tests/test_views.py::test_view_of_view[np_array-slice_subset-slice_subset]
0.07s call     anndata/tests/test_inplace_subset.py::test_inplace_subset_var[scipy_csr-array_bool_subset]
0.07s call     anndata/tests/test_inplace_subset.py::test_inplace_subset_obs[scipy_csr-array_subset]
0.07s call     anndata/tests/test_inplace_subset.py::test_inplace_subset_obs[np_array-array_subset]
0.07s call     anndata/tests/test_inplace_subset.py::test_inplace_subset_var[np_array-array_subset]
0.07s call     anndata/tests/test_inplace_subset.py::test_inplace_subset_obs[scipy_csr-array_int_subset]
0.07s call     anndata/tests/test_base.py::test_concatenate_dense
0.07s call     anndata/tests/test_inplace_subset.py::test_inplace_subset_var[scipy_csc-slice_subset]
0.06s call     anndata/tests/test_inplace_subset.py::test_inplace_subset_obs[scipy_csr-array_bool_subset]
0.06s call     anndata/tests/test_inplace_subset.py::test_inplace_subset_var[scipy_csc-array_subset]
0.06s call     anndata/tests/test_helpers.py::test_assert_equal_raw
0.06s call     anndata/tests/test_inplace_subset.py::test_inplace_subset_obs[scipy_csc-array_bool_subset]
0.06s call     anndata/tests/test_inplace_subset.py::test_inplace_subset_var[scipy_csc-array_bool_subset]
0.06s call     anndata/tests/test_inplace_subset.py::test_inplace_subset_var[scipy_csr-array_int_subset]
0.06s call     anndata/tests/test_inplace_subset.py::test_inplace_subset_obs[scipy_csc-array_subset]
0.06s call     anndata/tests/test_inplace_subset.py::test_inplace_subset_obs[scipy_csc-array_int_subset]
0.06s call     anndata/tests/test_inplace_subset.py::test_inplace_subset_obs[np_array-array_bool_subset]
0.06s call     anndata/tests/test_inplace_subset.py::test_inplace_subset_var[np_array-slice_subset]
0.06s call     anndata/tests/test_inplace_subset.py::test_inplace_subset_var[np_array-array_int_subset]
0.06s call     anndata/tests/test_inplace_subset.py::test_inplace_subset_obs[np_array-single_subset]
0.06s call     anndata/tests/test_inplace_subset.py::test_inplace_subset_var[scipy_csr-array_subset]
0.06s call     anndata/tests/test_inplace_subset.py::test_inplace_subset_var[scipy_csr-single_subset]
0.06s call     anndata/tests/test_inplace_subset.py::test_inplace_subset_var[scipy_csc-single_subset]
0.06s call     anndata/tests/test_inplace_subset.py::test_inplace_subset_obs[scipy_csc-single_subset]
0.06s call     anndata/tests/test_inplace_subset.py::test_inplace_subset_obs[np_array-slice_subset]
0.06s call     anndata/tests/test_inplace_subset.py::test_inplace_subset_obs[scipy_csr-single_subset]
0.06s call     anndata/tests/test_inplace_subset.py::test_inplace_subset_obs[scipy_csr-slice_subset]
0.06s call     anndata/tests/test_readwrite.py::test_readwrite_backed[typ0]
0.06s call     anndata/tests/test_inplace_subset.py::test_inplace_subset_var[scipy_csr-slice_subset]
0.06s call     anndata/tests/test_inplace_subset.py::test_inplace_subset_var[np_array-array_bool_subset]
0.06s call     anndata/tests/test_readwrite.py::test_readwrite_backed[csr_matrix]
0.06s call     anndata/tests/test_inplace_subset.py::test_inplace_subset_var[np_array-single_subset]
0.05s call     anndata/tests/test_readwrite.py::test_readwrite_loom[varm_names0-obsm_names0-typ0]
0.05s call     anndata/tests/test_readwrite.py::test_readwrite_loom[varm_names0-obsm_names0-csr_matrix]
0.05s call     anndata/tests/test_readwrite.py::test_readwrite_loom[varm_names1-obsm_names0-csr_matrix]
0.05s call     anndata/tests/test_readwrite.py::test_readwrite_loom[varm_names0-obsm_names1-csr_matrix]
0.05s call     anndata/tests/test_readwrite.py::test_readwrite_loom[varm_names1-obsm_names1-csr_matrix]
0.05s call     anndata/tests/test_readwrite.py::test_readwrite_loom[varm_names0-obsm_names1-typ0]
0.05s call     anndata/tests/test_readwrite.py::test_readwrite_loom[varm_names1-obsm_names0-typ0]
0.05s call     anndata/tests/test_readwrite.py::test_readwrite_loom[varm_names1-obsm_names1-typ0]
0.04s call     anndata/tests/test_readwrite.py::test_readwrite_h5ad_one_dimension[csr_matrix]
0.04s call     anndata/tests/test_readwrite.py::test_readwrite_h5ad_one_dimension[typ0]
0.04s call     anndata/tests/test_base.py::test_concatenate_dense_duplicates
0.04s call     anndata/tests/test_layers.py::test_readwrite_loom
0.04s call     anndata/tests/test_readwrite.py::test_scanpy_krumsiek11[zarr]
0.04s call     anndata/tests/test_views.py::test_double_index[array_subset-array_int_subset]
0.04s call     anndata/tests/test_views.py::test_double_index[array_subset-array_subset]
0.04s call     anndata/tests/test_views.py::test_double_index[array_bool_subset-array_subset]
0.04s call     anndata/tests/test_base.py::test_concatenate_mixed
0.04s call     anndata/tests/test_views.py::test_double_index[array_subset-array_bool_subset]
0.04s call     anndata/tests/test_views.py::test_double_index[array_int_subset-array_int_subset]
0.04s call     anndata/tests/test_views.py::test_double_index[array_subset-single_subset]
0.03s call     anndata/tests/test_views.py::test_double_index[array_bool_subset-single_subset]
0.03s call     anndata/tests/test_views.py::test_double_index[array_int_subset-array_bool_subset]
0.03s call     anndata/tests/test_base.py::test_boolean_slicing
0.03s call     anndata/tests/test_views.py::test_double_index[slice_subset-array_int_subset]
0.03s call     anndata/tests/test_base.py::test_concatenate_sparse
0.03s call     anndata/tests/test_views.py::test_double_index[array_bool_subset-slice_subset]
0.03s call     anndata/tests/test_views.py::test_double_index[slice_subset-array_subset]
0.03s call     anndata/tests/test_views.py::test_double_index[array_subset-slice_subset]
0.03s call     anndata/tests/test_hdf5_backing.py::test_backed_modification_sparse[csr_matrix]
0.03s call     anndata/tests/test_views.py::test_double_index[array_bool_subset-array_int_subset]
0.03s call     anndata/tests/test_views.py::test_double_index[array_bool_subset-array_bool_subset]
0.03s call     anndata/tests/test_views.py::test_double_index[slice_subset-slice_subset]
0.03s call     anndata/tests/test_views.py::test_double_index[single_subset-array_subset]
0.03s call     anndata/tests/test_views.py::test_double_index[array_int_subset-slice_subset]
0.03s call     anndata/tests/test_raw.py::test_raw
0.03s call     anndata/tests/test_views.py::test_double_index[slice_subset-single_subset]
0.03s call     anndata/tests/test_views.py::test_double_index[array_int_subset-array_subset]
0.03s call     anndata/tests/test_hdf5_backing.py::test_backed_modification_sparse[csc_matrix]
0.03s call     anndata/tests/test_views.py::test_view_of_view[scipy_csc-single_subset-array_bool_subset]
0.03s call     anndata/tests/test_views.py::test_double_index[slice_subset-array_bool_subset]
0.03s call     anndata/tests/test_views.py::test_double_index[single_subset-array_bool_subset]
0.03s call     anndata/tests/test_views.py::test_double_index[single_subset-single_subset]
0.03s call     anndata/tests/test_views.py::test_double_index[array_int_subset-single_subset]
0.03s call     anndata/tests/test_views.py::test_view_of_view[scipy_csc-single_subset-array_int_subset]
0.03s call     anndata/tests/test_views.py::test_double_index[single_subset-array_int_subset]
0.03s call     anndata/tests/test_views.py::test_double_index[single_subset-slice_subset]
0.03s call     anndata/tests/test_views.py::test_view_of_view[np_array-single_subset-array_bool_subset]
0.03s call     anndata/tests/test_hdf5_backing.py::test_double_index
0.03s call     anndata/tests/test_hdf5_backing.py::test_read_write_X[scipy-csc-r-False]
0.03s call     anndata/tests/test_views.py::test_view_of_view[scipy_csr-single_subset-slice_subset]
0.03s call     anndata/tests/test_views.py::test_view_of_view[scipy_csc-single_subset-slice_subset]
0.03s call     anndata/tests/test_readwrite.py::test_scanpy_krumsiek11[h5ad]
0.03s call     anndata/tests/test_views.py::test_view_of_view[scipy_csr-single_subset-array_subset]
0.03s call     anndata/tests/test_views.py::test_view_of_view[scipy_csr-single_subset-array_bool_subset]
0.03s call     anndata/tests/test_views.py::test_view_of_view[scipy_csc-single_subset-single_subset]
0.03s call     anndata/tests/test_views.py::test_view_of_view[np_array-single_subset-single_subset]
0.03s call     anndata/tests/test_views.py::test_view_of_view[scipy_csr-single_subset-array_int_subset]
0.03s call     anndata/tests/test_hdf5_backing.py::test_read_write_X[scipy-csr-r+-True]
0.03s call     anndata/tests/test_views.py::test_view_of_view[np_array-single_subset-array_int_subset]
0.03s call     anndata/tests/test_views.py::test_view_of_view[scipy_csr-single_subset-single_subset]
0.03s call     anndata/tests/test_views.py::test_view_of_view[np_array-single_subset-slice_subset]
0.03s call     anndata/tests/test_views.py::test_view_of_view[scipy_csc-single_subset-array_subset]
0.03s call     anndata/tests/test_views.py::test_view_of_view[np_array-single_subset-array_subset]
0.03s call     anndata/tests/test_hdf5_backing.py::test_return_to_memory_mode
0.03s call     anndata/tests/test_hdf5_backing.py::test_read_write_X[scipy-csr-r+-False]
0.03s call     anndata/tests/test_hdf5_backing.py::test_read_write_X[scipy-csc-r-True]
0.03s call     anndata/tests/test_hdf5_backing.py::test_read_write_X[scipy-csc-r+-False]
0.03s call     anndata/tests/test_hdf5_backing.py::test_read_write_X[scipy-csc-False-False]
0.03s call     anndata/tests/test_readwrite.py::test_write_categorical
0.02s call     anndata/tests/test_hdf5_backing.py::test_backed_modification
0.02s call     anndata/tests/test_hdf5_backing.py::test_read_write_X[scipy-csr-r-True]
0.02s call     anndata/tests/test_hdf5_backing.py::test_read_write_X[scipy-csc-r+-True]
0.02s call     anndata/tests/test_hdf5_backing.py::test_read_write_X[scipy-csc-False-True]
0.02s call     anndata/tests/test_hdf5_backing.py::test_read_write_X[scipy-csr-r-False]
0.02s call     anndata/tests/test_hdf5_backing.py::test_read_write_X[scipy-csr-False-False]
0.02s call     anndata/tests/test_hdf5_backing.py::test_read_write_X[scipy-csr-False-True]
0.02s call     anndata/tests/test_base.py::test_1d_slice_dtypes
0.02s call     anndata/tests/test_hdf5_backing.py::test_read_write_X[np-array-r-True]
0.02s call     anndata/tests/test_hdf5_backing.py::test_read_write_X[np-array-r+-True]
0.02s call     anndata/tests/test_views.py::test_set_varm
0.02s call     anndata/tests/test_hdf5_backing.py::test_read_write_X[np-array-r+-False]
0.02s call     anndata/tests/test_views.py::test_set_obsm
0.02s call     anndata/tests/test_hdf5_backing.py::test_read_write_X[np-array-r-False]
0.02s call     anndata/tests/test_hdf5_backing.py::test_read_write_X[np-array-False-True]
0.02s call     anndata/tests/test_readwrite.py::test_readwrite_hdf5_empty[read_zarr-write_zarr-test_empty.zarr]
0.02s call     anndata/tests/test_readwrite.py::test_read_excel
0.02s call     anndata/tests/test_hdf5_backing.py::test_read_write_X[np-array-False-False]
0.02s call     anndata/tests/test_h5py.py::test_create_and_read_dataset
0.02s setup    anndata/tests/test_h5py.py::test_create_and_read_dataset
0.02s call     anndata/tests/test_views.py::test_layers_view
0.01s call     anndata/tests/test_readwrite.py::test_readwrite_maintain_X_dtype[csr_matrix]
0.01s call     anndata/tests/test_readwrite.py::test_write_csv[csr_matrix]
0.01s call     anndata/tests/test_readwrite.py::test_write_csv[typ0]
0.01s call     anndata/tests/test_readwrite.py::test_readwrite_maintain_X_dtype[typ0]
0.01s call     anndata/tests/test_obsmvarm.py::test_setting_sparse
0.01s call     anndata/tests/test_obspvarp.py::test_setting_sparse
0.01s call     anndata/tests/test_readwrite.py::test_readwrite_sparse_as_dense
0.01s call     anndata/tests/test_views.py::test_set_subset_varm[array_int_subset]
0.01s call     anndata/tests/test_readwrite.py::test_readwrite_hdf5_empty[read_h5ad-write_h5ad-test_empty.h5ad]
0.01s call     anndata/tests/test_layers.py::test_readwrite
0.01s call     anndata/tests/test_views.py::test_set_var[array_bool_subset]
0.01s call     anndata/tests/test_h5py.py::test_dataset_append
0.01s call     anndata/tests/test_views.py::test_not_set_subset_X[scipy_csr-array_subset]
0.01s call     anndata/tests/test_views.py::test_set_var[slice_subset]
0.01s call     anndata/tests/test_views.py::test_set_obs[array_subset]
0.01s call     anndata/tests/test_views.py::test_not_set_subset_X[scipy_csr-array_int_subset]
0.01s call     anndata/tests/test_views.py::test_set_var[array_int_subset]
0.01s call     anndata/tests/test_views.py::test_modify_view_component[scipy_csc-varm]
0.01s call     anndata/tests/test_views.py::test_not_set_subset_X[scipy_csc-array_bool_subset]
0.01s call     anndata/tests/test_views.py::test_modify_view_component[scipy_csr-varm]
0.01s call     anndata/tests/test_views.py::test_not_set_subset_X[scipy_csc-slice_subset]
0.01s call     anndata/tests/test_views.py::test_not_set_subset_X[scipy_csr-array_bool_subset]
0.01s call     anndata/tests/test_views.py::test_not_set_subset_X[scipy_csr-slice_subset]
0.01s call     anndata/tests/test_views.py::test_set_subset_obsm[array_subset]
0.01s call     anndata/tests/test_views.py::test_modify_view_component[scipy_csr-layers]
0.01s call     anndata/tests/test_views.py::test_set_obs[array_bool_subset]
0.01s call     anndata/tests/test_views.py::test_set_var[array_subset]
0.01s call     anndata/tests/test_views.py::test_set_obs[array_int_subset]
0.01s call     anndata/tests/test_base.py::test_slicing
0.01s call     anndata/tests/test_views.py::test_not_set_subset_X[scipy_csc-array_int_subset]
0.01s call     anndata/tests/test_views.py::test_set_obs[slice_subset]
0.01s call     anndata/tests/test_views.py::test_not_set_subset_X[scipy_csc-array_subset]
0.01s call     anndata/tests/test_views.py::test_not_set_subset_X[scipy_csr-single_subset]
0.01s call     anndata/tests/test_h5py.py::test_create_dataset_from_dataset
0.01s call     anndata/tests/test_views.py::test_set_subset_obsm[single_subset]
0.01s call     anndata/tests/test_views.py::test_set_subset_obsm[slice_subset]
0.01s call     anndata/tests/test_views.py::test_set_var[single_subset]
0.01s call     anndata/tests/test_views.py::test_set_obs[single_subset]
0.01s call     anndata/tests/test_views.py::test_not_set_subset_X[scipy_csc-single_subset]
0.01s call     anndata/tests/test_views.py::test_modify_view_component[scipy_csc-layers]
0.01s call     anndata/tests/test_obspvarp.py::test_setting_ndarray
0.01s call     anndata/tests/test_views.py::test_modify_view_component[np_array-varm]
0.01s call     anndata/tests/test_views.py::test_modify_view_component[np_array-layers]
0.01s call     anndata/tests/test_views.py::test_modify_view_component[scipy_csr-obsm]
0.01s call     anndata/tests/test_views.py::test_not_set_subset_X[np_array-single_subset]
0.01s call     anndata/tests/test_views.py::test_set_subset_obsm[array_int_subset]
0.01s call     anndata/tests/test_views.py::test_modify_view_component[scipy_csc-obsm]
0.01s call     anndata/tests/test_views.py::test_set_subset_obsm[array_bool_subset]
0.01s call     anndata/tests/test_views.py::test_set_subset_varm[array_subset]
0.01s call     anndata/tests/test_views.py::test_modify_view_component[np_array-obsm]
0.01s call     anndata/tests/test_views.py::test_set_subset_varm[array_bool_subset]
0.01s call     anndata/tests/test_views.py::test_not_set_subset_X[np_array-array_subset]
0.01s call     anndata/tests/test_views.py::test_not_set_subset_X[np_array-slice_subset]
0.01s call     anndata/tests/test_views.py::test_set_subset_varm[single_subset]
0.01s call     anndata/tests/test_views.py::test_not_set_subset_X[np_array-array_int_subset]
0.01s call     anndata/tests/test_views.py::test_set_subset_varm[slice_subset]
0.01s call     anndata/tests/test_base.py::test_creation
0.01s call     anndata/tests/test_obsmvarm.py::test_setting_ndarray
0.01s call     anndata/tests/test_readwrite.py::test_readwrite_hdf5_empty[read_zarr-write_zarr-test_empty.zip]
0.01s call     anndata/tests/test_base.py::test_convenience
0.01s call     anndata/tests/test_views.py::test_views
0.01s call     anndata/tests/test_views.py::test_set_scalar_subset_X[scipy_csc-array_subset]
0.01s call     anndata/tests/test_views.py::test_set_varm_key
0.01s call     anndata/tests/test_views.py::test_not_set_subset_X[np_array-array_bool_subset]
0.01s call     anndata/tests/test_views.py::test_set_obsm_key
0.01s call     anndata/tests/test_base.py::test_slicing_strings
0.01s call     anndata/tests/test_views.py::test_set_scalar_subset_X[scipy_csc-single_subset]
0.01s call     anndata/tests/test_base.py::test_slicing_series
0.01s call     anndata/tests/test_obsmvarm.py::test_setting_dataframe

```
</details>

Reply:
That makes them much slower than all the other tests. But the full test suite is 40s after adding them, so going 35→40 isn’t that bad.

Reply:
I've halved the test time for these by reducing number of samples. Still the longest tests, but they do rely on a lot of data being written.